### PR TITLE
To change WarmUp Scheduler with ConstantLR and LinearLR

### PR DESCRIPTION
**Pull request by iramazanli**
Partially unblocks https://github.com/pytorch/vision/issues/4281

Previously we have added WarmUp Schedulers to PyTorch Core in the PR : https://github.com/pytorch/pytorch/pull/60836 which had two mode of execution - linear and constant depending on warming up function.

In this PR we are changing this interface to more direct form, as separating linear and constant modes to separate Schedulers. In particular 

```Python
scheduler1 = WarmUpLR(optimizer, warmup_factor=0.1, warmup_iters=5, warmup_method="constant")
scheduler2 = WarmUpLR(optimizer, warmup_factor=0.1, warmup_iters=5, warmup_method="linear")
```

will look like

```Python
scheduler1 = ConstantLR(optimizer, warmup_factor=0.1, warmup_iters=5)
scheduler2 = LinearLR(optimizer, warmup_factor=0.1, warmup_iters=5)
```

correspondingly.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -427,25 +427,78 @@ def _get_closed_form_lr(self):
...
+class ConstantLR(_LRScheduler):
+    """Decays the learning rate of each parameter group by a small constant factor until the
+    number of epoch reaches a pre-defined milestone: total_iters. Notice that such decay can
+    happen simultaneously with other changes to the learning rate from outside this scheduler.
     When last_epoch=-1, sets initial lr as lr.
 
     Args:
         optimizer (Optimizer): Wrapped optimizer.
-        warmup_factor (float): The number we multiply learning rate in the first epoch.
+        factor (float): The number we multiply learning rate in the first epoch.
```
Comment (datumbox) [inline torch/optim/lr_scheduler.py:438]: Since everywhere we talk about "decaying", does it make sense to assert that this value is always between [0,1]?

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -427,25 +427,78 @@ def _get_closed_form_lr(self):
...
+
+class LinearLR(_LRScheduler):
+    """Decays the learning rate of each parameter group by linearly increasing small
+    multiplicative factor until the number of epoch reaches a pre-defined milestone: total_iters.
+    Notice that such decay can happen simultaneously with other changes to the learning rate
+    from outside this scheduler. When last_epoch=-1, sets initial lr as lr.
+
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        start_value (float): The number we multiply learning rate in the first epoch.
```
Comment (datumbox) [inline torch/optim/lr_scheduler.py:495]: Perhaps rename it to `start_factor`, similar to Constant's param? Also assert it's between [0,1]

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -427,25 +427,78 @@ def _get_closed_form_lr(self):
...
+            The multiplication factor changes towards end_value in the following epochs.
+            Default: 1./3.
+        end_value (float): The number we multiply learning rate at the end of linear changing
```
Comment (datumbox) [inline torch/optim/lr_scheduler.py:498]: similarly `end_factor`? Also assert it's between [0,1]

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -427,25 +427,78 @@ def _get_closed_form_lr(self):
...
+        if (self.last_epoch == self.total_iters):
+            return [group['lr'] * (1.0 / self.factor) for group in self.optimizer.param_groups]
+
+    def _get_closed_form_lr(self):
+        return [base_lr * (self.factor + (self.last_epoch >= self.total_iters) * (1 - self.factor))
+                for base_lr in self.base_lrs]
+
+
+class LinearLR(_LRScheduler):
+    """Decays the learning rate of each parameter group by linearly increasing small
```
Comment (datumbox) [inline torch/optim/lr_scheduler.py:488]:
```suggestion
    """Decays the learning rate of each parameter group by linearly changing small
```

File: torch/optim/lr_scheduler.py [hunk 5]
```diff
@@ -427,25 +427,78 @@ def _get_closed_form_lr(self):
                 for base_lr in self.base_lrs]
 
 
-class WarmUpLR(_LRScheduler):
-    """Decays the learning rate of each parameter group by either a small constant
-    or linearly increasing small warmup factor until the number of epoch reaches a
-    pre-defined milestone: warmup_iters. Notice that such decay can happen
-    simultaneously with other changes to the learning rate from outside this scheduler.
+class ConstantLR(_LRScheduler):
```
Comment (datumbox) [inline torch/optim/lr_scheduler.py:430]: We need to review the descriptions of this class because in some parts it still talks about warming up and "if the method is constant".

File: torch/optim/lr_scheduler.py [hunk 6]
```diff
@@ -457,50 +510,38 @@ class WarmUpLR(_LRScheduler):
         >>> # lr = 0.0375   if epoch == 2
         >>> # lr = 0.04375  if epoch == 3
         >>> # lr = 0.005    if epoch >= 4
-        >>> scheduler = WarmUpLR(self.opt, warmup_factor=0.5, warmup_iters=4, warmup_method="linear")
+        >>> scheduler = LinearLR(self.opt, start_value=0.5, total_iters=4)
```
Comment (datumbox) [inline torch/optim/lr_scheduler.py:513]: I agree leaving the end param out of the example. It's very rare that end != 1.0 and can lead to strange behaviours.

**Review by datumbox — COMMENTED** [review]
@iramazanli Thanks a lot for the fast PR. 

I left a few minor comments for which I would like your input, but nothing major.

I also noticed that we don't have any tests for LinearLR that use the new end parameter. We need to adapt the tests to include it I think. Other than this it looks good.

File: torch/optim/lr_scheduler.py (hunk 5 above)
Comment (iramazanli) [reply torch/optim/lr_scheduler.py:430]: thanks for pointing it out, i noticed it before and fixed in local version. but it seems that i haven't updated the remote version before.

File: torch/optim/lr_scheduler.py (hunk 4 above)
Comment (iramazanli) [reply torch/optim/lr_scheduler.py:488]: good point.

**Review by datumbox — APPROVED** [review]
LGTM, thanks for the speedy changes.

File: torch/optim/lr_scheduler.py [hunk 7]
```diff
@@ -457,50 +510,44 @@ class WarmUpLR(_LRScheduler):
...
                           "please use `get_last_lr()`.", UserWarning)
 
         if self.last_epoch == 0:
-            return [group['lr'] * self.warmup_factor for group in self.optimizer.param_groups]
+            return [group['lr'] * self.start_factor for group in self.optimizer.param_groups]
 
-        if (self.last_epoch > self.warmup_iters or
-                (self.warmup_method == "constant" and self.last_epoch != self.warmup_iters)):
+        if (self.last_epoch > self.total_iters):
             return [group['lr'] for group in self.optimizer.param_groups]
```
Comment (fmassa) [inline torch/optim/lr_scheduler.py:542]: Don't we need to multiply by `end_factor` now that we might have an `end_factor` which is different than 1?

File: torch/optim/lr_scheduler.py [hunk 8]
```diff
@@ -457,50 +510,44 @@ class WarmUpLR(_LRScheduler):
...
 
-        if (self.warmup_method == "constant" and self.last_epoch == self.warmup_iters):
-            return [group['lr'] * (1.0 / self.warmup_factor) for group in self.optimizer.param_groups]
-
-        return [group['lr'] * (1. + (1.0 - self.warmup_factor) /
-                (self.warmup_iters * self.warmup_factor + (self.last_epoch - 1) * (1 - self.warmup_factor)))
+        return [group['lr'] * (1. + (self.end_factor - self.start_factor) /
+                (self.total_iters * self.start_factor + (self.last_epoch - 1) * (self.end_factor - self.start_factor)))
```
Comment (fmassa) [inline torch/optim/lr_scheduler.py:545]: Just asking to make sure this case is handled properly: can we have `end_factor < start_factor`  (i.e., for a linear decay of the learning rate)?

**Review by fmassa — COMMENTED** [review]
Thanks a lot for the PR Ilqar!

I've made a couple of minor comments for my understanding, otherwise everything LGTM!

File: torch/optim/lr_scheduler.py (hunk 8 above)
Comment (iramazanli) [reply torch/optim/lr_scheduler.py:545]: yes it can have :)

File: torch/optim/lr_scheduler.py (hunk 7 above)
Comment (iramazanli) [reply torch/optim/lr_scheduler.py:542]: Actually as this is the recursive definition and at the epoch number total_iters we already have multiplication with end_factor, we no more need to mention multiplication here.
